### PR TITLE
fix(#21): batch docwrites to minimize time spent on POST calls

### LINKED ISSUE #21: Support uploading documents in bulk
I have a design with 10 district hospitals, each with 50 health centers, and 20 households per health center. It looks like documents are generated and uploaded separately. I've been waiting for at least 20 minutes for 30k docs to be generated and uploaded. 

We should save documents before uploading them, and upload them in bulk to save time.

### PR DESCRIPTION
**Pull request by jkuester**
Closes https://github.com/medic/test-data-generator/issues/21

I did some testing locally to try and profile the performance here.  Firstly, running locally on my laptop (with 32GB of RAM and 12 CPU threads), pushing any large amount of data  (20,000+ docs) into `medic` would quickly bottleneck on view indexing, maxing out the CPU, _regardless of how the doc writes were batched_. This bottlenecking affects the test-data-generator runs because it seems like Couch's ability to process new `_bulk_docs` requests is severely impacted by having a bunch of indexing jobs already running. Eventually the TDG process starts to hang on the POST calls to `_bulk_docs` because Couch is not responding.  I suspect it may just be a matter of thread-pool exhaustion.  Whatever the case, the changes in this PR are unlikely to significantly improve observed write speeds to _local CouchDB instances_.  Writing (even smaller amounts of data) to remote CouchDb instances will be faster, though, since this change does optimize the network requests.  

To get a better idea of raw data write speeds, I also tested against an empty CouchDB (with no views).  This way I was able to better measure the data throughput without bottlenecking on the view indexing.  Against this empty database I did a very unscientific comparison of TDG runtime for writing 200,000 docs (from a non-trivial hierarchical design) with different batch sizes:

| Batch Size | Runtime |
| --- | --- |
| `master` | 67sec |
| 500 | 37sec |
| 1000 | 32sec |
| 2000 | 32sec |
| 5000 | 34sec |

I ended up settling on a batch size of `1000` docs. It is a nice round number, and anything higher did not provide any clear benefit.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,23 @@ Based on that previous example. The tool will create and push 14 report docs at
...
+
+### Generate then replicate
+
+When generating a massive dataset (100,000+ documents) for an existing database with views (e.g. the `medic` database), it may be preferable to actually generate the data into a temporary database (on the same CouchDB instance) and then replicate the data into the target database (e.g. using Fauxton). The document creation rate for this process (including generating into the temp db, replicating to `medic` db, and indexing views) has been measured at `~11,000 docs/min`.
+
+While the speed gains of this approach are somewhat modest, it has a couple additional benefits:
+
+- The time spent running the actual generation script is minimal (since docs are added to the temp db at `~360,000 docs/min`). So, there is less danger of the script being interrupted or becoming disconnected from the CouchDB instance. Once all the docs are saved in the temp db, the replication/indexing happen internally within the Couch instance. This is where the vast majority of the time is spent.
+- You can keep your generated dataset in the temp db for future use, without having to regenerate it each time you want to test with it.
```
Comment (latin-panda) [inline README.md]:
```suggestion
When generating a massive dataset (100,000+ documents) for an existing database with views (e.g. the `medic` database), it may be preferable to actually generate the data into a temporary database (on the same CouchDB instance) and then replicate the data into the target database (e.g. using Fauxton). The document creation rate for this process (including generating into the temp DB, replicating to `medic` DB, and indexing views) has been measured at `~11,000 docs/min`.

While the speed gains of this approach are somewhat modest, it has a couple additional benefits:

- The time spent running the actual generation script is minimal (since docs are added to the temp db at `~360,000 docs/min`). So, there is less danger of the script being interrupted or becoming disconnected from the CouchDB instance. Once all the docs are saved in the temp DB, the replication/indexing happens internally within the Couch instance. This is where the vast majority of the time is spent.
- You can keep your generated dataset in the temp DB for future use, without having to regenerate it each time you want to test with it.
```

Comment (latin-panda) [reply README.md]: > and then replicate the data into the target database (e.g. using Fauxton)

Would you happen to have any link that gives info on manually doing that replication in Fauxton?

**Review by latin-panda — APPROVED** [review]
Thanks a lot for fixing this Josh! ✨

After merging it, I'm going to move the documentation to CHT Docs

File: README.md (hunk 1 above)
Comment (jkuester) [reply README.md]: > Would you happen to have any link that gives info on manually doing that replication in Fauxton?

Unfortunately, all the online doc for Fauxton appear to be very outdated (and not super useful to link to).  The good news is that replication in Fauxton is very straightforward. I included a relative link to the Fauxton form for setting up a replication job, and it is very easy to get everything filled out.
